src/Database/DateInterval.php

```diff
@@ -0,0 +1,158 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Database;
+
+use UnexpectedValueException;
+
+/**
+ * TODO: Add documentation
+ */
+class DateInterval extends \DateInterval
```

Comment:
I didn't see this used in the tests anywhere, do we still need it?

Reply:
Added tests and missing documentation.

Reply:
Also wanted to make sure: Do you think this class will be confusing for people? If so, I can rename it. I don't see any classes with the same name as their base and wanted to make sure there wasn't a requirement against that.

Reply:
I wasn't sure why we needed this interval class. I couldn't figure that out from the tests/code.

Reply:
It seems to duplicate some of the behavior in ChronosInterval. If folks need the additional precision beyond what PHP native objects provide they could use that package.

Reply:
I moved the `\DateInterval`-related code out of the `IntervalExpression` to keep it separate, so it's more of a utility class. It takes the `\DateInterval` properties (y, m, d, h, etc) and maps them into the database-specific units (YEAR, MONTH, DAY, HOUR, etc).

For comparison, `ChronosInterval` is used to reverse a `\DateInterval` object back into a period specification string `P1Y10D5H`.

This class instead creates an array `[ 'YEAR' => 1, 'DAY' => 10, 'HOUR' => 5 ]` that is used by `IntervalExpression` to generate `INTERVAL '1-0' YEAR_MONTH + INTERVAL '10 05:00:00.00' DAY_MICROSECOND` or the more specific `INTERVAL 1 YEAR + INTERVAL 10 DAY + INTERVAL 5 HOUR`. It also does some checks such as ensuring there is at least one valid value specified (creating an object with `P0Y` would not produce any change in a date it was applied to) as well as preventing relative or special formats (`second tuesday of march`) as those do not produce usable properties and are not supported by most databases. Since the methods are specific to database I left it under the `Cake\Database` namespace.